### ISSUE TITLE
Different cache folder for different envs

### ISSUE THREAD

**Issue by fasterthanlime** [electronjs/electron-compilers#76]
*This issue applies both to electron-compile and electron-compilers, sorry*

Reading through a bunch of code right now to find out if a different `$TEMP/compileCache_${hash}` will get generated if `NODE_ENV` changes, and.. it doesn't like it does?

Here's the default cache directory code:

```javascript
export function calculateDefaultCompileCacheDirectory() {
  let tmpDir = process.env.TEMP || process.env.TMPDIR || '/tmp';
  let hash = require('crypto').createHash('md5').update(process.execPath).digest('hex');

  let cacheDir = path.join(tmpDir, `compileCache_${hash}`);
  mkdirp.sync(cacheDir);

  d(`Using default cache directory: ${cacheDir}`);
  return cacheDir;
}
```

The problem is:

  * One might want to enable code coverage (or other babel plugins) only in testing
  * But since all files are cached by their content
  * And the files wouldn't change, only the compiler settings would
  * Then electron-compile would just happily serve the old cached versions (which may have coverage instrumentation when they shouldn't, and the other way around).

You can probably reproduce the problem easily if you use electron-compile with a `.babelrc` like:

```json
{
  "env": {
    "production": {
      "presets": ["es2016-node5", "react"],
      "sourceMaps": false
    },
    "development": {
      "presets": ["es2016-node5", "react"],
      "sourceMaps": "inline"
    }
  }
}
```

run it once with `NODE_ENV=production`, then again with `NODE_ENV=development`, I'm 80% confident you'd end up not seeing sourceMaps still.

---

Here's my proposal for electron-compile:

  * [x] Change `calculateDefaultCompileCacheDirectory` to feed off of
    * process.execPath
    * process.env.NODE_ENV || "development"
    * the digest for the info/config/opts object (using `digest-for-object`)
    * see https://github.com/fasterthanlime/electron-compile/commit/a072bafdf2b7655af35acde4ce6880fbb2da305d

Tangentially, here's what I've got planned for electron-compilers:

  * [x] Add way for compilers to get an input sourceMap
    * compilerContext seems okay
  * [x] Make babel compiler consume that input sourceMap when it's there
    * see https://github.com/fasterthanlime/electron-compilers/commit/a98839a4e0263dfbc9ad1e8eff794aef2ca7f77c
  * [x] Drop sorcery dependency
    * see https://github.com/fasterthanlime/electron-compilers/commit/a98839a4e0263dfbc9ad1e8eff794aef2ca7f77c
  * [x] Make both `coverage` options (typescript & babel) only apply when `NODE_ENV=test`
    * Alternatively: only do env matching when `coverage` is a string, and `coverage: true` would *always* instrument, so that the old behavior is retained?
    * Alternatively: don't change anything about the coverage option, force folks to use env blocks
    * see https://github.com/fasterthanlime/electron-compilers/commit/a98839a4e0263dfbc9ad1e8eff794aef2ca7f77c
  * [x] Make both `coverage` option use [babel-plugin-istanbul](https://github.com/istanbuljs/babel-plugin-istanbul) (a relatively thin wrapper over new-istanbul's instrumenting lib).
    * For babel, this means pushing adding a plug-in to the plugins list
    * For typescript, this.. requires babel :/ 
    * see https://github.com/fasterthanlime/electron-compilers/commit/a98839a4e0263dfbc9ad1e8eff794aef2ca7f77c
  * [x] Drop dependency to istanbul 0.4.5 (legacy istanbul) - it's somehow still in electron-compile's package.json
    * see https://github.com/fasterthanlime/electron-compile/commit/e7c2ddebd0b293e73dca67ac731bdf48d1a9c3f6
  * [x] Document how to set up typescript+babel (babel block in the typescript config)
    * see https://github.com/fasterthanlime/electron-compile/commit/1f7a795fb64ad718c18db3ed14caab0a9db34bf1
  * [x] Document how to set up coverage instrumentation
    * see https://github.com/fasterthanlime/electron-compile/commit/1f7a795fb64ad718c18db3ed14caab0a9db34bf1
  * [x] Document what coverage instrumentation does (collects coverage info using the new istanbul, aka istanbuljs/istanbuljs, into the global `__coverage__` variable)
    * see https://github.com/fasterthanlime/electron-compile/commit/1f7a795fb64ad718c18db3ed14caab0a9db34bf1
  * [x] Give an example of collecting/reporting from a `__coverage__` variable, with proper sourceMap support (it's like 8 lines).
    * see https://github.com/fasterthanlime/electron-compile/commit/1f7a795fb64ad718c18db3ed14caab0a9db34bf1

Let me know if any of that sounds bad @paulcbetts & others so I can adjust plans accordingly!

**Comment by fasterthanlime** [comment]
Additional notes on istanbul@0.4.5 being deprecated: it only became clear to me after spending a day investigating, but: the acorn-based istanbul is pretty dead.

The new istanbul is heavily marketed as `nyc` (its command-line interface), it's pretty much everything <https://istanbul.js.org/> talks about - but there's actual libs underneath, conveniently stashed in [this monorepo](http://github.com/istanbuljs/istanbuljs/).

The new instrumenter is based on babel, which makes a bunch of sense:

  * If you're already using babel, then the cost of `istanbul-lib-instrument` is next to non-existent
  * If you're not, well.. at least it's more reliable than acorn was.

Quick size comparison:

### Legacy istanbul

```
istanbul-only ツ npm i --save istanbul

+ istanbul@0.4.5
added 60 packages in 5.097s
 istanbul-only ツ du -ch node_modules | tail -1
9.4M    total
```

### babel-plugin-istanbul

```
babel-plugin ツ npm i --save babel-plugin-istanbul

+ babel-plugin-istanbul@4.1.4
added 107 packages in 8.772s
 babel-plugin ツ du -ch node_modules | tail -1
18M     total
```

N.B: size of the instrument+coverage+babel plug-in only:

```
babel-plugin ツ du -ch node_modules/istanbul-lib-* node_modules/babel-plugin-istanbul | tail -1
164K    total
```

---

TL;DR istanbul is now babel-based, it kinda blows if you're not already using babel, but I don't think electron-compile{,rs} should keep using istanbul@0.4.5

**Comment by fasterthanlime** [comment]
All done in:

  * https://github.com/electron/electron-compilers/pull/77
  * https://github.com/electron/electron-compile/pull/247

Works great for me (typescript+babel, inline source maps):

![coverage-demo](https://user-images.githubusercontent.com/7998310/27382075-a9ee7988-5685-11e7-839d-050004f6b318.png)

**Comment by fasterthanlime** [comment]
Closing in favor of both PRs.
